rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+                Some(c) => c,
+            };
+            let axes = [
+                position.0.x - cursor.0.x,
+                position.0.y - cursor.0.y,
+                position.0.z - cursor.0.z,
+            ];
+
+            let mut negative_padding = 3;
+            let mut negative_axes = [0x00, 0x00, 0x00];
+            let mut adjusted_axes = [[0x00, 0x00, 0x00], [0x00, 0x00, 0x00], [0x00, 0x00, 0x00]];
+            for (key, axis) in axes.iter().enumerate() {
+                if *axis < 0 {
+                    negative_axes[key] = 0xFF;
+                }
+
+                let axis_filler = match axis.abs() {
+                    ..256 => 3,
+                    256..65536 => 2,
+                    65536.. => 1,
+                };
+                if axis_filler < negative_padding {
+                    negative_padding = axis_filler;
+                }
+
+                // FIXME: This is really ugly
+                let mut axis_adjuster = axis.abs();
+                while axis_adjuster > 0 {
+                    match axis_adjuster {
+                        ..256 => {
+                            adjusted_axes[2][key] = axis_adjuster as u8;
+                            axis_adjuster -= axis_adjuster;
+                        }
+                        256..65536 => {
+                            let offset = axis_adjuster / 256;
+                            adjusted_axes[1][key] += offset as u8;
+                            axis_adjuster -= offset * 256;
+                        }
+                        65536.. => {
+                            let offset = axis_adjuster / 65536;
+                            adjusted_axes[0][key] += offset as u8;
+                            axis_adjuster -= offset * 65536;
+                        }
+                    }
+                }
+            }
+
+            for _ in 0..negative_padding {
+                data.extend(negative_axes.iter())
+            }
+
+            // 3 -> 1, 2 -> 2, 1 -> 3. Amount of 256 multiples to write
+            for i in 0..(4 - negative_padding) {
+                data.extend(adjusted_axes[2 - i].iter());
+            }
+
+            data.extend(chunk.encode());
+            chunk_cursor = Some(position);
+        }
+
+        data
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn encode_default() {
+        let mut terr = Terrain::new();
+        let mut chunk = Chunk::new_with_base(Material::Air);
+        let mut voxel = Voxel::new_with_water(Material::Grass, 1.0, 0.5);
+        for m in 2..=22 {
+            voxel.set_material(Material::try_from(m as u8).unwrap());
+            chunk.write_voxel(&VoxelCoordinates::new(m - 2, 0, 0), voxel);
+        }
+        terr.write_chunk(&ChunkCoordinates::default(), chunk.clone());
+        terr.write_chunk(&ChunkCoordinates::new(1, 0, 0), chunk.clone());
+
+        let encoded = base64::encode(terr.encode());
+        println!("{}", encoded);
```

Comment:
...This test doesn't actually do anything. Can we make it panic if the output isn't identical to something Roblox emitted?

Reply:
this test is solely for development testing purposes while i was making the encoder so I could put it in place files and hence shouldn't make it out of the draft PR stage